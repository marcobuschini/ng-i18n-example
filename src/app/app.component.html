<mat-radio-group
    aria-label="{ Select a language | translate }"
    (change)="setLanguage($event.value)">
    <mat-radio-button
    *ngFor="let culture of cultures"
    [value]="culture.isoCode">
        {{ culture.name }}
    </mat-radio-button>
</mat-radio-group>

<mlb-parking-widget #widget [vendor]="vendor" (buying)="addToCart($event)"></mlb-parking-widget>

<h2>Shopping cart</h2>
<mat-list>
    <mat-list-item *ngFor="let item of items">
        {{ item.name }}
    </mat-list-item>
</mat-list>
