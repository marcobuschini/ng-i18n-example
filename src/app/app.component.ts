import { Component, ViewChild, OnInit, OnDestroy } from '@angular/core';

import { TranslateService, Culture, Translations } from 'ng-i18n';
import { of, Subscription } from 'rxjs';
import { HttpClient } from '@angular/common/http';

import { Feature, Vendor, ParkingSlot } from 'widget';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit, OnDestroy {
  public title = 'ng-i18n-example';
  public vendor: Vendor = {
    name: 'Test Vendor',
    description: 'This is a test vendor',
    features: [
      { name: 'feature 1' },
      { name: 'feature 2' },
      { name: 'feature 3' }
    ] as Feature[]
  };
  public items = new Array<ParkingSlot>();

  @ViewChild('widget', { static: true })
  public widget;

  public cultures: Culture[];

  private culture$: Subscription;

  public constructor(public translate: TranslateService, private http: HttpClient) {
    this.translate.addCulture(
      { isoCode: 'en_US', name: 'English (US)'} as Culture,
      {
        culture: { isoCode: 'en_US', name: 'English (US)'} as Culture,
        translations: this.http.get('assets/widget/i18n/en-US.js')
      } as Translations
    );
    this.translate.addCulture(
      { isoCode: 'it_IT', name: 'Italiano (Italia)'} as Culture,
      {
        culture: { isoCode: 'it_IT', name: 'Italiano (Italia)'} as Culture,
        translations: this.http.get('assets/widget/i18n/it-IT.js')
      } as Translations
    );
  }

  public ngOnInit() {
    this.culture$ = this.translate.getAvailableCultures().subscribe(
      success => {
        this.cultures = Array.from(success);
        this.translate.setCulture(this.cultures[0]);
        this.widget.vendor = this.vendor;
      }
    );
  }

  public ngOnDestroy() {
    this.culture$.unsubscribe();
  }

  public addToCart(slot: ParkingSlot) {
    this.items.push(slot);
  }

  public setLanguage(culture: Culture) {
    console.log(culture.name);
    this.translate.setCulture(culture);
  }
}
