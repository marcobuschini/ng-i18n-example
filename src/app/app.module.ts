import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

import { MatListModule } from '@angular/material/list';
import { MatRadioModule } from '@angular/material/radio';
import { HttpClientModule } from '@angular/common/http';

import { NgI18nModule } from 'ng-i18n';
import { WidgetModule } from 'widget';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    HttpClientModule,
    MatListModule,
    MatRadioModule,
    NgI18nModule,
    WidgetModule,
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
