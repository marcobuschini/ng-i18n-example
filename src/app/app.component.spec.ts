import { TestBed, ComponentFixture, fakeAsync, flush } from '@angular/core/testing';
import { AppComponent } from './app.component';
import { WidgetComponent, Feature, Vendor, ParkingSlot, WidgetModule } from 'widget';
import { MatCardModule } from '@angular/material/card';
import { MatDividerModule } from '@angular/material/divider';
import { MatListModule } from '@angular/material/list';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { TranslateTestingModule } from 'ngx-translate-testing';
import { of } from 'rxjs';
import { By } from '@angular/platform-browser';
import {
  TranslateModule,
  MissingTranslationHandler,
  MissingTranslationHandlerParams,
  TranslateFakeLoader,
  TranslateLoader
} from '@ngx-translate/core';

describe('AppComponent', () => {
  let fixture: ComponentFixture<AppComponent>;
  let component: AppComponent;
  let addSpy: jest.SpyInstance;

  const dummyVendor: Vendor = {
    name: 'Test Vendor',
    description: 'This is a test vendor',
    features: [
      { name: 'Feature 1' } as Feature,
      { name: 'Feature 2' } as Feature,
      { name: 'Feature 3' } as Feature
    ] as Feature[]
  };

  const dummySlots: ParkingSlot[] = [
    {
      id: 0,
      name: 'Slot 1',
      features: [
        'Feature 1',
        'Feature 2'
      ]
    }
  ];

  const ENGLISH_LANGUAGE = 'en-US';
  const ENGLISH_TRANSLATIONS = {
    'Add to cart': 'Add to cart',
    Collapse: 'Collapse',
    Expand: 'Expand'
};

  const ITALIAN_LANGUAGE = 'es';
  const ITALIAN_TRANSLATIONS = {
    'Add to cart': 'Aggiungi al carrello',
    Collapse: 'Riduci',
    Expand: 'Espandi'
  };

  const TRANSLATIONS = {
    [ENGLISH_LANGUAGE]: ENGLISH_TRANSLATIONS,
    [ITALIAN_LANGUAGE]: ITALIAN_TRANSLATIONS
  };

  beforeEach(async () => {
    TestBed.configureTestingModule({
      declarations: [
        AppComponent,
      ],
      imports: [
        MatCardModule,
        MatDividerModule,
        MatListModule,
        HttpClientTestingModule,
        WidgetModule,
        TranslateTestingModule.withTranslations(TRANSLATIONS)
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(AppComponent);
    fixture.componentInstance.widget = TestBed.createComponent(WidgetComponent).componentInstance;

    const vendorSpy = jest.spyOn(fixture.componentInstance.widget.service, 'getVendorFeatures')
      .mockReturnValue(of(dummyVendor.features));
    const slotsSpy = jest.spyOn(fixture.componentInstance.widget.service, 'getParkingSlots')
      .mockReturnValue(of(dummySlots));
    addSpy = jest.spyOn(fixture.componentInstance, 'addToCart');

    component = fixture.componentInstance;
    component.widget.ngOnInit();
    fixture.detectChanges();
  });

  afterEach(() => {
    component.widget.ngOnDestroy();
  });

  it('should create the application, and add an item to the cart', fakeAsync(() => {
    expect(component).toBeTruthy();

    flush();

    expect(fixture).toMatchSnapshot();

    const button = fixture.debugElement.queryAll(By.css('mlb-parking-widget button'))[0].nativeElement as HTMLButtonElement;
    button.click();
    fixture.detectChanges();

    expect(addSpy).toHaveBeenCalled();

    expect(fixture).toMatchSnapshot();
  }));
});
