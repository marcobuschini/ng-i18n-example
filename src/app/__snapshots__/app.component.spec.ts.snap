// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the application, and add an item to the cart 1`] = `
<app-root
  addToCart={[Function Function]}
  items={[Function Array]}
  title={[Function String]}
  vendor={[Function Object]}
  widget={[Function WidgetComponent]}
>
  <mlb-parking-widget
    _nghost-a-c0=""
    ng-reflect-vendor="[object Object]"
  >
    <mat-card
      _ngcontent-a-c0=""
      class="mat-card"
    >
      <mat-card-title
        _ngcontent-a-c0=""
        class="mat-card-title"
      >
        Test Vendor
      </mat-card-title>
      <mat-card-subtitle
        _ngcontent-a-c0=""
        class="mat-card-subtitle"
      >
        This is a test vendor
      </mat-card-subtitle>
      <mat-card-content
        _ngcontent-a-c0=""
        class="mat-card-content"
      >
        <mat-list
          _ngcontent-a-c0=""
          class="mat-list mat-list-base"
          role="list"
        >
          
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 1 
            </div>
          </mat-list-item>
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 2 
            </div>
          </mat-list-item>
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 3 
            </div>
          </mat-list-item>
        </mat-list>
        <mat-divider
          _ngcontent-a-c0=""
          aria-orientation="horizontal"
          class="mat-divider mat-divider-horizontal"
          role="separator"
        />
        <div
          _ngcontent-a-c0=""
          fxlayout.gt-xs="row"
          fxlayout.lt-sm="column"
          fxlayoutalign="space-between"
          ng-reflect-fx-layout-align="space-between"
          ng-reflect-fx-layout.gt-xs="row"
          ng-reflect-fx-layout.lt-sm="column"
          style="align-content: stretch; align-items: stretch; box-sizing: border-box; display: flex; flex-direction: row; justify-content: space-between;"
        >
          
          <div
            _ngcontent-a-c0=""
          >
             Slot 1 
            <button
              _ngcontent-a-c0=""
              class="mat-stroked-button mat-button-base"
              mat-stroked-button=""
            >
              <span
                class="mat-button-wrapper"
              >
                Add to cart
              </span>
              <div
                class="mat-button-ripple mat-ripple"
                matripple=""
                ng-reflect-centered="false"
                ng-reflect-disabled="false"
                ng-reflect-trigger="[object HTMLButtonElement]"
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            
            <button
              _ngcontent-a-c0=""
              class="mat-stroked-button mat-button-base"
              mat-stroked-button=""
            >
              <span
                class="mat-button-wrapper"
              >
                Expand
              </span>
              <div
                class="mat-button-ripple mat-ripple"
                matripple=""
                ng-reflect-centered="false"
                ng-reflect-disabled="false"
                ng-reflect-trigger="[object HTMLButtonElement]"
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            
            
          </div>
        </div>
      </mat-card-content>
    </mat-card>
  </mlb-parking-widget><h2>
    Shopping cart
  </h2><mat-list
    class="mat-list mat-list-base"
  >
    
  </mat-list>
</app-root>
`;

exports[`AppComponent should create the application, and add an item to the cart 2`] = `
<app-root
  addToCart={[Function Function]}
  items={[Function Array]}
  title={[Function String]}
  vendor={[Function Object]}
  widget={[Function WidgetComponent]}
>
  <mlb-parking-widget
    _nghost-a-c0=""
    ng-reflect-vendor="[object Object]"
  >
    <mat-card
      _ngcontent-a-c0=""
      class="mat-card"
    >
      <mat-card-title
        _ngcontent-a-c0=""
        class="mat-card-title"
      >
        Test Vendor
      </mat-card-title>
      <mat-card-subtitle
        _ngcontent-a-c0=""
        class="mat-card-subtitle"
      >
        This is a test vendor
      </mat-card-subtitle>
      <mat-card-content
        _ngcontent-a-c0=""
        class="mat-card-content"
      >
        <mat-list
          _ngcontent-a-c0=""
          class="mat-list mat-list-base"
          role="list"
        >
          
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 1 
            </div>
          </mat-list-item>
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 2 
            </div>
          </mat-list-item>
          <mat-list-item
            _ngcontent-a-c0=""
            class="mat-list-item"
            role="listitem"
          >
            <div
              class="mat-list-item-content"
            >
              <div
                class="mat-list-item-ripple mat-ripple"
                mat-ripple=""
                ng-reflect-disabled="true"
                ng-reflect-trigger="[object HTMLUnknownElement]"
              />
              <div
                class="mat-list-text"
              />
              Feature 3 
            </div>
          </mat-list-item>
        </mat-list>
        <mat-divider
          _ngcontent-a-c0=""
          aria-orientation="horizontal"
          class="mat-divider mat-divider-horizontal"
          role="separator"
        />
        <div
          _ngcontent-a-c0=""
          fxlayout.gt-xs="row"
          fxlayout.lt-sm="column"
          fxlayoutalign="space-between"
          ng-reflect-fx-layout-align="space-between"
          ng-reflect-fx-layout.gt-xs="row"
          ng-reflect-fx-layout.lt-sm="column"
          style="align-content: stretch; align-items: stretch; box-sizing: border-box; display: flex; flex-direction: row; justify-content: space-between;"
        >
          
          <div
            _ngcontent-a-c0=""
          >
             Slot 1 
            <button
              _ngcontent-a-c0=""
              class="mat-stroked-button mat-button-base"
              mat-stroked-button=""
            >
              <span
                class="mat-button-wrapper"
              >
                Add to cart
              </span>
              <div
                class="mat-button-ripple mat-ripple"
                matripple=""
                ng-reflect-centered="false"
                ng-reflect-disabled="false"
                ng-reflect-trigger="[object HTMLButtonElement]"
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            
            <button
              _ngcontent-a-c0=""
              class="mat-stroked-button mat-button-base"
              mat-stroked-button=""
            >
              <span
                class="mat-button-wrapper"
              >
                Expand
              </span>
              <div
                class="mat-button-ripple mat-ripple"
                matripple=""
                ng-reflect-centered="false"
                ng-reflect-disabled="false"
                ng-reflect-trigger="[object HTMLButtonElement]"
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
            
            
          </div>
        </div>
      </mat-card-content>
    </mat-card>
  </mlb-parking-widget><h2>
    Shopping cart
  </h2><mat-list
    class="mat-list mat-list-base"
  >
    
    <mat-list-item
      class="mat-list-item"
    >
      <div
        class="mat-list-item-content"
      >
        <div
          class="mat-list-item-ripple mat-ripple"
          mat-ripple=""
          ng-reflect-disabled="true"
          ng-reflect-trigger="[object HTMLUnknownElement]"
        />
        <div
          class="mat-list-text"
        />
         Slot 1 
      </div>
    </mat-list-item>
  </mat-list>
</app-root>
`;
